File: common-core/src/main/java/io/quartic/common/client/WebsocketListener.java
package io.quartic.common.client;

import com.fasterxml.jackson.databind.JavaType;
import io.quartic.common.serdes.ObjectMappersKt;
import org.immutables.value.Value;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import rx.Observable;

import javax.websocket.DeploymentException;
import javax.websocket.Endpoint;
import javax.websocket.EndpointConfig;
import javax.websocket.Session;
import java.io.IOException;
import java.net.URISyntaxException;

import static io.quartic.common.serdes.ObjectMappersKt.objectMapper;
import static rx.Observable.empty;
import static rx.Observable.just;

@Value.Immutable
public abstract class WebsocketListener<T> {
    private static final Logger LOG = LoggerFactory.getLogger(WebsocketListener.class);

    @Value.Immutable
    public static abstract class Factory {
        public static Factory of(String url, WebsocketClientSessionFactory websocketFactory) {
            return ImmutableFactory.of(url, websocketFactory);
        }

        @Value.Parameter
        protected abstract String url();

        @Value.Parameter
        protected abstract WebsocketClientSessionFactory websocketFactory();

        public <T> WebsocketListener<T> create(Class<T> type) {
            return create(objectMapper().getTypeFactory().uncheckedSimpleType(type));
        }

        public <T> WebsocketListener<T> create(JavaType type) {
            return ImmutableWebsocketListener.<T>builder().type(type)
                    .url(url())
                    .websocketFactory(websocketFactory())
                    .build();
        }
    }

    protected abstract JavaType type();
    protected abstract String url();
    protected abstract WebsocketClientSessionFactory websocketFactory();

    @Value.Lazy
    public Observable<T> observable() {
        return Observable.<String>create(sub -> {
            final Endpoint endpoint = new Endpoint() {
                @Override
                public void onOpen(Session session, EndpointConfig config) {
                    session.addMessageHandler(String.class, sub::onNext);
                }
            };

            try {
                websocketFactory().create(endpoint, url());
            } catch (URISyntaxException | DeploymentException | IOException e) {
                sub.onError(e);
            }
        }).flatMap(this::convert).share();
    }


    private Observable<T> convert(String message) {
        try {
            return just(objectMapper().readValue(message, type()));
        } catch (IOException e) {
            LOG.error("Error converting message", e);
            return empty();
        }
    }

}

File: common-core/src/main/kotlin/io/quartic/common/client/WebsocketListener.kt
package io.quartic.common.client

import com.fasterxml.jackson.databind.JavaType
import io.quartic.common.logging.logger
import io.quartic.common.serdes.OBJECT_MAPPER
import io.quartic.common.serdes.objectMapper
import rx.Emitter.BackpressureMode
import rx.Observable
import rx.Observable.*
import java.io.IOException
import javax.websocket.Endpoint
import javax.websocket.EndpointConfig
import javax.websocket.MessageHandler.Whole
import javax.websocket.Session

class WebsocketListener<T>(
        private val type: JavaType,
        private val url: String,
        private val websocketFactory: WebsocketClientSessionFactory
) {
    private val LOG by logger()

    class Factory(private val url: String, private val websocketFactory: WebsocketClientSessionFactory) {
        fun <T> create(type: Class<T>): WebsocketListener<T> {
            return create(OBJECT_MAPPER.typeFactory.uncheckedSimpleType(type))
        }

        fun <T> create(type: JavaType): WebsocketListener<T> {
            return WebsocketListener(type = type, url = url, websocketFactory = websocketFactory)
        }
    }

    val observable: Observable<T> by lazy {
        emitter.flatMap({ this.convert(it) }).share()
    }

    private val emitter: Observable<String>
        get() {
            return fromEmitter({ sub ->
                val endpoint = object : Endpoint() {
                    override fun onOpen(session: Session, config: EndpointConfig) {
                        session.addMessageHandler(Whole<String> { sub.onNext(it) })
                    }
                }

                try {
                    websocketFactory.create(endpoint, url)
                } catch (e: Exception) {
                    sub.onError(e)
                }
            }, BackpressureMode.BUFFER)
        }


    private fun convert(message: String) = try {
        just(objectMapper().readValue<T>(message, type))
    } catch (e: IOException) {
        LOG.error("Error converting message", e)
        empty<T>()
    }
}
